### ISSUE TITLE
Server capable of decrypting some messages

### ISSUE THREAD

**Issue by wbl** [PeerioTechnologies/peerio-client#333]
In crypto.js we see that decryptAuthTokens decrypts a server-provided message with the user's private key. But the public key that sent the message is also provided by the server, and can be set by the server to anything. The only reason this doesn't immediately break the system is the check for the first two bytes and the total length of the message. I've not yet found a way to break the encryption of files or messages using this attack, but one may exist. Any value 32 bytes in length starting with the two bytes 'AT' may be decrypted if sent to a user, and it seems to be only luck that prevents exploitation.

My recommendation is that server authentication of clients use a different key from the one with which messages are encrypted.

**Comment by anri-asaturov** [comment]
Hi, thank you, I'm looking into it.
In case you didn't know, we have a bug bounty programme https://peerio.zendesk.com/hc/en-us/articles/203981385-Bug-Bounty

**Comment by anri-asaturov** [comment]
Seems that you are right, also we've found that this was reported by cure53 at the first audit round and 'fixed' by famous whistleblower Nadim Kobeissi: https://github.com/PeerioTechnologies/peerio-legacy/issues/216 

After verifying this with auditors I think we are going to do what you suggested, derive another key and use it for server challenges.

Thanks again!

**Comment by anri-asaturov** [comment]
Just noticed that repo is private, you can look at the issue here: https://dl.dropboxusercontent.com/u/7081215/peerio/PT-01-007%20Client%20decrypts%20arbitrary%20ciphertexts%20for%20the%20server%20%28Critical%29%20%C2%B7%20Issue%20%23216%20%C2%B7%20PeerioTechnologies_peerio-legacy.html

**Comment by flohdot** [comment]
We took a look at this problem and asked cure53 to dig in as well. 

We found that the only plaintext that did not have format restrictions was read receipts. Other plaintexts were either JSON-encoded (messages/files) or had a prefix requirement (tokens). 

However, as per cure53's report on the matter:

> Tricking another client into treating a non-receipt as a receipt only has very limited utility: The client will perform a string encoding conversion on the received data, slice the last 13 characters away and compare the rest against a received receipt string. Depending on the result, a message will be shown as verified or unverified to the local user, but the result is not transmitted over the network.

That said, the utility of the read receipt proofs that we have provided thus far is limited, and we have a better design in mind that both avoids this particular issue and is also more performant. This will be implemented in the coming months. In the meantime, in our latest release(s), we have removed read receipt en/decryption to mitigate any possible issues. 

We are in the process of expanding/rewriting our documentation and we'll include a piece on plaintext formats for clarity in this rewrite.
